src/c++/perf_analyzer/test_request_rate_manager.cc

```diff
@@ -312,6 +312,104 @@ class TestRequestRateManager : public TestLoadManagerBase,
     StopWorkerThreads();
   }
 
+  /// Helper function that will setup and run a case to verify custom data
+  /// behavior
+  /// \param tensors Vector of input ModelTensors
+  /// \param json_str The custom data json text
+  /// \param expected_values Vector of expected input values for each inference
+  /// \param expect_init_failure True if InitManager is expected to throw an
+  /// error
+  /// \param expect_thread_failure True if the thread is expected to have
+  /// an error
+  void TestCustomData(
+      std::vector<ModelTensor>& tensors, const std::string json_str,
+      std::vector<std::vector<int32_t>>& expected_values,
+      bool expect_init_failure, bool expect_thread_failure)
+  {
+    params_.user_data = {json_str};
+
+    TestCustomDataSetup(tensors);
+
+    if (expect_init_failure) {
+      REQUIRE_THROWS_AS(
+          InitManager(
+              params_.string_length, params_.string_data, params_.zero_input,
+              params_.user_data, params_.start_sequence_id,
+              params_.sequence_id_range, params_.sequence_length,
+              params_.sequence_length_specified,
+              params_.sequence_length_variation),
+          PerfAnalyzerException);
+      return;
+    } else {
+      REQUIRE_NOTHROW(InitManager(
+          params_.string_length, params_.string_data, params_.zero_input,
+          params_.user_data, params_.start_sequence_id,
+          params_.sequence_id_range, params_.sequence_length,
+          params_.sequence_length_specified,
+          params_.sequence_length_variation));
+    }
+
+    auto thread_status = TestCustomDataRunThread();
+
+    if (!thread_status.IsOk() && !expect_thread_failure) {
+      std::cout << "Unexpected thread failure: " << thread_status.Message()
+                << std::endl;
+    }
+
+    REQUIRE(thread_status.IsOk() != expect_thread_failure);
+    if (!thread_status.IsOk()) {
+      return;
```

Comment:
This isn't an error?

Reply:
if thread status is not ok and we do NOT expect that to happen, it will fail on line above.  
if thread status is not ok and we do expect that to happen, then we return here because it did what we expected and the rest of the information is bogus.

Reply:
I recoded this section for clarity